ACTIN-2067: Add 'name' to PrimaryTumor data model and remove location/type/details

ninajacobs opened the pull request:
WIP

jbartletthmf reviewed (commented):
Nice refactor 🙂

ninajacobs commented on clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/emc/extraction/TumorDetailsExtractor.kt:
```diff
@@ -79,7 +79,8 @@ class TumorDetailsExtractor(
 
-        val hasBrainOrGliomaTumor = primaryTumorDetails.primaryTumorLocation == "Brain" || primaryTumorDetails.primaryTumorType == "Glioma"
+        val hasBrainOrGliomaTumor =
+            primaryTumorDetails.name.lowercase().contains("brain") || primaryTumorDetails.name.lowercase().contains("glioma")
```
I checked the standard extractor too, I think what we did there is not correct. We won't have "brain" anyway (since Dutch) and the dutch word for brain can be wrong (can be brain metastases). I changed that a bit now too

ninajacobs commented on algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/tumor/HasLocallyAdvancedCancer.kt:
```diff
@@ -4,25 +4,26 @@ import com.hartwig.actin.algo.evaluation.EvaluationFactory
-            EvaluationFactory.fail("Stage $stage is not considered locally advanced")
-        }
-    }
+        val stageIIIEvaluationResult = HasTumorStage(setOf(TumorStage.III)).evaluate(record).result
+        val stageIIEvaluationResult = HasTumorStage(setOf(TumorStage.II)).evaluate(record).result
```
Thanks, I didn't see that correctly and mis-corrected it I think. Would you review if it is now okay?

jbartletthmf commented on clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/standard/extraction/StandardTumorDetailsExtractor.kt:
```diff
@@ -50,22 +50,23 @@ class StandardTumorDetailsExtractor(
         with(ehrPatientRecord.tumorDetails) {
-            val hasBrainOrGliomaTumor =
-                tumorType?.lowercase()?.contains("glioma") == true || tumorLocation?.lowercase()?.contains("brain") == true
+            val hasGliomaTumor =
+                listOf("glioma").any { tumorType?.lowercase()?.contains(it) == true || tumorLocation?.lowercase()?.contains(it) == true }
```
```suggestion
            val hasGliomaTumor = listOfNotNull(tumorType, tumorLocation).any { it.lowercase().contains("glioma") }
```

jbartletthmf commented on algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/tumor/HasLocallyAdvancedCancer.kt:
```diff
@@ -10,20 +10,20 @@ import com.hartwig.actin.datamodel.clinical.TumorStage
     override fun evaluate(record: PatientRecord): Evaluation {
+        val stage = record.tumor.stage ?: return EvaluationFactory.undetermined("Locally advanced cancer undetermined (tumor stage missing)")
         val stageIIIEvaluationResult = HasTumorStage(setOf(TumorStage.III)).evaluate(record).result
         val stageIIEvaluationResult = HasTumorStage(setOf(TumorStage.II)).evaluate(record).result
```
The fact that this evaluation function is wrapped by the derived stage evaluator means we shouldn't have to call functions that use derived stages internally. We should be able to drop these.

The derived evaluator will find the derived stages for this patient in the case where the real stage is not known, and evaluate this function for all derived stages. If there's a unique derived stage, it essentially works the same as if the stage was known. If there are multiple derived stages, it will combine the evaluation results for all of them according to `DerivedTumorStageEvaluationFunction`. For example, if the rule passes for one derived stage but not all, it will return `UNDETERMINED`.

Is there any reason we can't use the derived stage evaluator here? If there is, we could drop its use in the mapper and continue to use the `HasTumorStage` evaluations.

jbartletthmf commented on algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/tumor/HasLocallyAdvancedCancer.kt:
```diff
@@ -10,20 +10,20 @@ import com.hartwig.actin.datamodel.clinical.TumorStage
-        val stageMessage = record.tumor.stage?.display() ?: "(derived stage: ${record.tumor.derivedStages?.joinToString(" or ")})"
 
         return when {
             stageIIIEvaluationResult == EvaluationResult.PASS -> {
```
```suggestion
            (stage.category ?: stage) == TumorStage.III -> {
```

jbartletthmf commented on algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/tumor/HasLocallyAdvancedCancer.kt:
```diff
@@ -10,20 +10,20 @@ import com.hartwig.actin.datamodel.clinical.TumorStage
+                EvaluationFactory.pass("Stage $stage is considered locally advanced")
             }
 
             stageIIEvaluationResult != EvaluationResult.FAIL || stageIIIEvaluationResult != EvaluationResult.FAIL -> {
```
```suggestion
            (stage.category ?: stage) == TumorStage.II -> {
```

ninajacobs commented:
Hi John, so if I understand/understood correctly:
- DerivedTumorStageEvaluationFunction handles all rules in case of derived stages. Hence none of the rules have to handle derived stage logic (except for stage messaging - I couldn't find code for that, so I did make a stageMessage for all rules)?
- Based on this, I refactored all the classes, and also added DerivedTumorStageEvaluationFunction for HasTumorStage.

Hope I now interpreted correctly!

jbartletthmf commented on algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/tumor/HasTumorStage.kt:
```diff
@@ -1,73 +1,31 @@
-import com.hartwig.actin.datamodel.algo.EvaluationResult
 import com.hartwig.actin.datamodel.clinical.TumorStage
 
 class HasTumorStage(private val stagesToMatch: Set<TumorStage>) : EvaluationFunction {
```
I'm not sure why this class didn't rely on the derived evaluation before (messages? testing?) but this seems like a nice refactor.

ninajacobs replied:
me neither..!

jbartletthmf commented on algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/tumor/HasUnresectableCancer.kt:
```diff
@@ -9,28 +9,23 @@ import com.hartwig.actin.datamodel.clinical.TumorStage
-        )
+        val stage =
+            record.tumor.stage ?: return EvaluationFactory.undetermined("Undetermined if cancer is unresectable (tumor stage missing)")
+        val stageMessage =
+            record.tumor.stage?.display() ?: "(derived stage: ${record.tumor.derivedStages?.joinToString(" or ") { it.display() }})"
```
`stage.display()`

jbartletthmf commented on algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/tumor/HasUnresectableCancer.kt:
```diff
@@ -9,28 +9,23 @@ import com.hartwig.actin.datamodel.clinical.TumorStage
     }
-
-    companion object {
-        private fun isStageMatch(stage: TumorStage, stageToMatch: TumorStage): Boolean {
-            return stage == stageToMatch || stage.category == stageToMatch
-        }
```
I don't mind using this function in place of the `stage.category ?: stage` evaluation used above. It might actually be a little more obvious and easy to read. We could move this to a helper object/`TumorStage` for reuse so we don't have to remember to check the category everywhere.

jbartletthmf reviewed (commented):
I like how much of the tests we can clean up when we rely on the derived stage evaluator.

jbartletthmf commented:
> * DerivedTumorStageEvaluationFunction handles all rules in case of derived stages. Hence none of the rules have to handle derived stage logic (except for stage messaging - I couldn't find code for that, so I did make a stageMessage for all rules)?

In `DerivedTumorStageEvaluationFunction`:
```kotlin
        val derivedResults = record.tumor.derivedStages?.associateWith { tumorStage -> evaluatedDerivedStage(record, tumorStage) }
        // ...
    }

    private fun evaluatedDerivedStage(record: PatientRecord, newStage: TumorStage): Evaluation {
        return originalFunction.evaluate(
            record.copy(tumor = record.tumor.copy(stage = newStage))
        )
    }
```
The derived stage evaluator evaluates the function for each derived stage assuming that the real stage is the derived stage (`record.copy(tumor = record.tumor.copy(stage = newStage))`), so there's no need to consider the derived stages at all in each function. Does that make sense?
